File: feign-core/src/main/java/name/abhijitsarkar/feign/core/service/IgnorablePropertiesMerger.java
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * License for more details.
 */

package name.abhijitsarkar.feign.core.service;

import lombok.extern.slf4j.Slf4j;
import name.abhijitsarkar.feign.core.model.FeignProperties;
import name.abhijitsarkar.feign.model.AbstractIgnorableRequestProperties;
import name.abhijitsarkar.feign.model.RequestProperties;
import org.springframework.util.ReflectionUtils;

import java.beans.IntrospectionException;
import java.beans.Introspector;
import java.beans.PropertyDescriptor;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Stream;

import static org.springframework.util.ReflectionUtils.invokeMethod;

/**
 * @author Abhijit Sarkar
 */
@Slf4j
@SuppressWarnings({"PMD.ConfusingTernary", "PMD.NullAssignment"})
public class IgnorablePropertiesMerger {
    public void merge(RequestProperties requestProperties, FeignProperties feignProperties) {
        List<PropertyDescriptor> globalIgnoreProperties =
                Arrays.asList(getPropertyDescriptors(AbstractIgnorableRequestProperties.class));

        Stream.of(
                requestProperties.getPath(),
                requestProperties.getMethod(),
                requestProperties.getQueries(),
                requestProperties.getHeaders(),
                requestProperties.getBody())
                .filter(Objects::nonNull)
                .forEach(requestProperty -> {
                    Class<?> clazz = requestProperty.getClass();
                    log.debug("Introspecting class: {}.", clazz.getName());

                    Arrays.stream(getPropertyDescriptors(clazz))
                            .forEach(property -> {
                                String name = property.getName();

                                globalIgnoreProperties.stream()
                                        .filter(globalProperty -> !"class".equals(name)
                                                && globalProperty.getName().equals(name))
                                        .findFirst()
                                        .ifPresent(globalProperty -> {
                                            log.debug("Merging property: {}.", name);

                                            Method getter = property.getReadMethod();
                                            Method superGetter = globalProperty.getReadMethod();

                                            Object ignore = null;

                                            if (getter != null) {
                                                ignore = invokeMethod(getter, requestProperty);
                                            } else if (superGetter != null) {
                                                ignore = invokeMethod(superGetter, requestProperty);
                                            }

                                            if (ignore == null) {
                                                Object globalIgnore = (superGetter != null) ?
                                                        invokeMethod(superGetter, feignProperties) : null;
                                                Method setter = property.getWriteMethod();
                                                Method superSetter = globalProperty.getWriteMethod();

                                                if (setter != null) {
                                                    invokeMethod(setter, requestProperty, globalIgnore);
                                                } else if (superSetter != null) {
                                                    invokeMethod(superSetter, requestProperty, globalIgnore);
                                                }
                                            }
                                        });
                            });
                });
    }

    private PropertyDescriptor[] getPropertyDescriptors(Class<?> clazz) {
        try {
            return Introspector.getBeanInfo(clazz).getPropertyDescriptors();
        } catch (IntrospectionException e) {
            ReflectionUtils.rethrowRuntimeException(e);
        }

        return new PropertyDescriptor[]{};
    }
}

File: feign-core/src/main/scala/name/abhijitsarkar/feign/core/service/IgnorablePropertiesMerger.scala
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of the GNU General Public License accompanies this software,
 * and is also available at http://www.gnu.org/licenses.
 */

package name.abhijitsarkar.feign.core.service

import java.beans.Introspector
import java.lang.reflect.Method
import java.lang.{Boolean => JavaBoolean}

import name.abhijitsarkar.feign.core.model.{AbstractIgnorableRequestProperties, FeignProperties, RequestProperties}
import org.slf4j.LoggerFactory

/**
  * @author Abhijit Sarkar
  */
class IgnorablePropertiesMerger {
  val logger = LoggerFactory.getLogger(classOf[IgnorablePropertiesMerger])

  def merge(requestProperties: RequestProperties, feignProperties: FeignProperties) {
    val globalPd = getPropertyDescriptors(classOf[AbstractIgnorableRequestProperties])

    List(requestProperties.path,
      requestProperties.method,
      requestProperties.queries,
      requestProperties.headers,
      requestProperties.body
    )
      .foreach { requestProperty =>
        logger.debug(s"Introspecting request property: ${requestProperty}.")
        val pd = getPropertyDescriptors(requestProperty.getClass)

        pd
          .zip(Stream.continually(globalPd))
          .flatMap(x => {
            val z = x._2
              .find(y => y.getName != "class" && y.getName == x._1.getName)

            List((Option(x._1), z))
          })
          .foreach {
            _ match {
              case (Some(local), Some(global)) => {
                val getter = local.getReadMethod
                val superGetter = global.getReadMethod

                Option(getter)
                  .flatMap(invokeGetter(_, requestProperty))
                  .orElse(invokeSuperGetter(superGetter, feignProperties))
                  .foreach { ignore =>
                    val setter = local.getWriteMethod

                    logger.debug(s"Setting ${local.getName} to $ignore.")

                    Option(setter)
                      .foreach(_.invoke(requestProperty, ignore))
                  }
              }
              case _ =>
            }
          }
      }
  }

  def getPropertyDescriptors(clazz: Class[_]) = Introspector.getBeanInfo(clazz).getPropertyDescriptors.toList

  def invokeGetter(getter: Method, requestProperty: Object) = {
    val ignore = Option(getter).flatMap(x => {
      val g = x.invoke(requestProperty).asInstanceOf[JavaBoolean]

      if (g == null) None else Some(g)
    })

    logger.debug(s"Local getter returned: ${ignore}.")

    ignore
  }

  def invokeSuperGetter(superGetter: Method, feignProperties: FeignProperties) = {
    val ignore = Option(superGetter).flatMap(x => {
      val g = x.invoke(feignProperties).asInstanceOf[JavaBoolean]

      if (g == null) None else Some(g)
    })

    logger.debug(s"Super getter returned: ${ignore}.")

    ignore
  }
}
